src/object/srv_obj_migrate.c

```diff
@@ -1775,77 +1804,98 @@ migrate_one_destroy(struct migrate_one *mrone)
+		tls->mpt_rmg = rmg;
+	} else {
+		D_ASSERTF(tls->mpt_rmg == rmg, "target=%d, rmg_off=%d\n", dmi->dmi_tgt_id,
+			  (int)(tls->mpt_rmg - &migr_eng_res.er_rmgs[0]));
+	}
 
-		if (yielded)
-			*yielded = true;
-		ABT_mutex_lock(tls->mpt_inflight_mutex);
-		ABT_cond_wait(tls->mpt_inflight_cond, tls->mpt_inflight_mutex);
-		ABT_mutex_unlock(tls->mpt_inflight_mutex);
-		if (tls->mpt_fini)
-			D_GOTO(out, rc = migrate_pool_tls_get_status(tls));
+	res = &rmg->rmg_resources[res_type];
+	while (1) {
+		if (tls->mpt_fini) {
+			rc = migrate_pool_tls_get_status(tls);
+			D_GOTO(out, rc);
+		}
+
+		if (res->res_units + units <= res->res_limit) {
+			res->res_units += units;
+			break;
+		}
 
-		ult_cnt = migrate_tgt_ult_cnt(tls, ult_type);
+		ABT_mutex_lock(rmg->rmg_mutex);
+		res->res_waiters++;
+		if (res->res_waiters >= 100 && res->res_waiters % 100 == 0) {
+			D_DEBUG(DB_REBUILD, "%d waiters are waiting on res=%s (target=%d)\n",
+				res->res_waiters, res->res_name, dmi->dmi_tgt_id);
+		}
+
+		ABT_cond_wait(res->res_cond, rmg->rmg_mutex);
+		res->res_waiters--;
+		ABT_mutex_unlock(rmg->rmg_mutex);
+		waited = true;
 	}
+	if (yielded)
+		*yielded = waited;
 
-	if (ult_type == OBJ_ULT)
+	/* per-pool counters for rebuild status tracking */
+	if (res_type == MIGR_OBJ)
 		tls->mpt_tgt_obj_ult_cnt++;
-	else
+	else if (res_type == MIGR_KEY)
 		tls->mpt_tgt_dkey_ult_cnt++;
+	else
+		tls->mpt_inflight_size += units;
+
+	D_DEBUG(DB_REBUILD, DF_RB " migrate %s: hold units=%lu, used=%lu, limit=%lu, waited=%d\n",
+		DP_RB_MPT(tls), res->res_name, units, res->res_units, res->res_limit, waited);
 out:
 	return rc;
 }
 
 static void
-migrate_tgt_try_wakeup(struct migrate_pool_tls *tls, int ult_type)
+migrate_res_release(struct migrate_pool_tls *tls, int res_type, long units)
 {
-	uint32_t ult_cnt = 0;
+	struct migr_res_manager *rmg;
+	struct migr_resource    *res;
 
-	ult_cnt = migrate_tgt_ult_cnt(tls, ult_type);
-	D_ASSERT(dss_get_module_info()->dmi_xs_id != 0);
-	if (tls->mpt_inflight_max_ult / 2 > ult_cnt) {
-		ABT_mutex_lock(tls->mpt_inflight_mutex);
-		ABT_cond_signal(tls->mpt_inflight_cond);
-		ABT_mutex_unlock(tls->mpt_inflight_mutex);
-	}
-}
+	rmg = tls->mpt_rmg;
+	D_ASSERT(rmg != NULL);
 
-static void
-migrate_tgt_exit(struct migrate_pool_tls *tls, int ult_type)
-{
-	D_ASSERT(dss_get_module_info()->dmi_xs_id != 0);
-	if (ult_type == OBJ_ULT) {
+	res = &rmg->rmg_resources[res_type];
+
+	D_DEBUG(DB_REBUILD, DF_RB "migrate %s: release units=%lu, used=%lu, limit=%lu\n",
+		DP_RB_MPT(tls), res->res_name, units, res->res_units, res->res_limit);
+
+	if (res_type == MIGR_OBJ) {
 		D_ASSERT(tls->mpt_tgt_obj_ult_cnt > 0);
 		tls->mpt_tgt_obj_ult_cnt--;
-	} else {
+	} else if (res_type == MIGR_KEY) {
 		D_ASSERT(tls->mpt_tgt_dkey_ult_cnt > 0);
 		tls->mpt_tgt_dkey_ult_cnt--;
+	} else {
+		D_ASSERT(tls->mpt_inflight_size >= units);
+		tls->mpt_inflight_size -= units;
+	}
+
+	D_ASSERT(res->res_units >= units);
+	res->res_units -= units;
+	if (res->res_waiters > 0) {
```

Comment:
should the test of res->res_waiters be done instead while the rmg->rmg_mutex lock is held? It seems like it should be evaluated while holding the lock.

Reply:
all ULTs changing res_waiters are on the same xstream so it's not required right? The only reason we have mutex is because ABT_cond requires it.